usr.bin/sockstat/sockstat.c

```diff
@@ -1107,235 +1093,420 @@ sctp_path_state(int state)
+					{ .socket = s->splice_socket });
+				if (sp != NULL)
+					formataddr(&sp->laddr->address,
+								buf, bufsize);
+			}
+			else
+				strlcpy(buf, "??", bufsize);
+			printf(" %-*s", cw->splice_address, buf);
+		}
+		if (opt_i) {
+			if (s->proto == IPPROTO_TCP ||
+				s->proto == IPPROTO_UDP) {
+				printf(" %*" PRIu64, cw->inp_gencnt ,s->inp_gencnt);
+			}
+			else
+				printf(" %*s", cw->inp_gencnt, "??");
+		}
+		if (opt_U) {
+			if (faddr != NULL &&
+				((s->proto == IPPROTO_SCTP &&
+					s->state != SCTP_CLOSED &&
+					s->state != SCTP_BOUND &&
+					s->state != SCTP_LISTEN) ||
+					(s->proto == IPPROTO_TCP &&
+					s->state != TCPS_CLOSED &&
+					s->state != TCPS_LISTEN))) {
+				printf(" %*u", cw->encaps, ntohs(faddr->encaps_port));
+			}
+			else
+				printf(" %*s", cw->encaps, "??");
+		}
+		if (opt_s) {
+			if (faddr != NULL &&
+				s->proto == IPPROTO_SCTP &&
+				s->state != SCTP_CLOSED &&
+				s->state != SCTP_BOUND &&
+				s->state != SCTP_LISTEN) {
+				printf(" %-*s", cw->path_state,
+					sctp_path_state(faddr->state));
+			}
+			else
+				printf(" %-*s", cw->path_state, "??");
+		}
+		if (first) {
+			if (opt_s) {
+				if (s->proto == IPPROTO_SCTP ||
+					s->proto == IPPROTO_TCP) {
+					switch (s->proto) {
+					case IPPROTO_SCTP:
+						printf(" %-*s", cw->conn_state,
+							sctp_conn_state(s->state));
+						break;
+					case IPPROTO_TCP:
+						if (s->state >= 0 &&
+							s->state < TCP_NSTATES)
+							printf(" %-*s", cw->conn_state,
+								tcpstates[s->state]);
 						else
-							pos += xprintf("?");
+							printf(" %-*s", cw->conn_state, "??");
 						break;
 					}
 				}
-				offset += 13;
+				else
+					printf(" %-*s", cw->conn_state, "??");
 			}
 			if (opt_S) {
-				if (s->proto == IPPROTO_TCP) {
-					do
-						pos += xprintf(" ");
-					while (pos < offset);
-					pos += xprintf("%.*s",
-					    TCP_FUNCTION_NAME_LEN_MAX,
-					    s->stack);
-				}
-				offset += TCP_FUNCTION_NAME_LEN_MAX + 1;
+				if (s->proto == IPPROTO_TCP)
+					printf(" %-*.*s", TCP_FUNCTION_NAME_LEN_MAX,
+						TCP_FUNCTION_NAME_LEN_MAX, s->stack);
+				else
+					printf(" %-*s", cw->stack, "??");
 			}
 			if (opt_C) {
-				if (s->proto == IPPROTO_TCP) {
-					do
-						pos += xprintf(" ");
-					while (pos < offset);
-					xprintf("%.*s", TCP_CA_NAME_MAX, s->cc);
-				}
-				offset += TCP_CA_NAME_MAX + 1;
+				if (s->proto == IPPROTO_TCP)
+					printf(" %-*.*s", TCP_CA_NAME_MAX,
+						TCP_CA_NAME_MAX, s->cc);
+				else
+					printf(" %-*s", cw->cc, "??");
 			}
 		}
 		if (laddr != NULL)
 			laddr = laddr->next;
 		if (faddr != NULL)
 			faddr = faddr->next;
-		if ((laddr != NULL) || (faddr != NULL)) {
-			xprintf("\n");
-			pos = 0;
-		}
-		first = 0;
+		if (laddr != NULL || faddr != NULL)
+			printf("%-*s %-*s %-*s %-*s %-*s",
+				cw->user,		"",
+				cw->command,	"",
+				cw->pid,		"",
+				cw->fd,			"",
+				cw->proto,		"");
+		first = false;
 	}
-	xprintf("\n");
+	printf("\n");
```

Comment:
Pro tip: using `printf` repeatedly is inefficient, because every call results in a `write(2)` syscall.  Using `fprintf` is much more efficient if you plan to call it over and over, because it buffers the output in the C library, and only calls `write(2)` when the buffer is full or you intentionally flush it.  HOWEVER, don't make that change now, because in a future commit you'll replace all of these with `xo_emit` instead.